src/controller/CHIPDeviceController.cpp

```diff
@@ -1770,13 +1770,13 @@ void DeviceCommissioner::CleanupCommissioning(DeviceProxy * proxy, NodeId nodeId
 
     if (completionStatus.err == CHIP_NO_ERROR)
     {
+        CommissioningStageComplete(CHIP_NO_ERROR);
         CommissioneeDeviceProxy * commissionee = FindCommissioneeDevice(nodeId);
         if (commissionee != nullptr)
         {
             ReleaseCommissioneeDevice(commissionee);
         }
         // Send the callbacks, we're done.
-        CommissioningStageComplete(CHIP_NO_ERROR);
```

Comment:
This does not looks right.  `CommissioningStageComplete` should mean we actually completed the step...

Reply:
@bzbarsky-apple 
This process is a function that is called only in the CleanUp Stage. Currently, the root cause of this issue is that mDeviceBeingCommissioned, which is used when CommissioningStageComplete is called, is already free. This free is done by ReleaseCommissioneeDevice.

Crash Point : 
https://github.com/project-chip/connectedhomeip/blob/824ba6dbd9a7e783353f4c714e7ca0ba03615c64/src/controller/CHIPDeviceController.cpp#L1903

Free Code :
https://github.com/project-chip/connectedhomeip/blob/824ba6dbd9a7e783353f4c714e7ca0ba03615c64/src/controller/CHIPDeviceController.cpp#L618


I think ReleaseCommissioneeDevice should be done at the end of the whole process. For this reason, the two codes have been changed, and they are implemented exactly the same in the code below, so this change was made. Please check again.

https://github.com/project-chip/connectedhomeip/blob/824ba6dbd9a7e783353f4c714e7ca0ba03615c64/src/controller/CHIPDeviceController.cpp#L1851

Reply:
Hmm.  I agree that if `CommissioningStageComplete` is going to assert `mDeviceBeingCommissioned` then it can't come after `ReleaseCommissioneeDevice`.  

So why is this not crashing on every single commissioning?

Reply:
@bzbarsky-apple 
If you do not set other optional parameters in the app, two DeviceProxy are created during the commissioning process. One is CommissioneeDeviceProxy and the other is OperationalDeviceProxy.
In the general case, mDeviceBeingCommissioned is set to OperationalDeviceProxy, and it is CommissioneeDeviceProxy that needs to be freed, so it is not set to null in the code below.
https://github.com/project-chip/connectedhomeip/blob/6a7961db7960dec75039047210c87154fa514f22/src/controller/CHIPDeviceController.cpp#L616

However, when an issue occurs, SkipCommissioningComplete is set to true. In this case, because CASE is being skipped, OperationalDeviceProxy is not created and ends by switching to the CleanUp step. For this reason, because it is the same value in the above code, it is set to null, and a crash occurs.

Reply:
So hold on a sec.  Are we then leaking the CommissioneeDeviceProxy when commissioning _does_ go through all the way to CommissioningComplete, because we've reset mDeviceBeingCommissioned to be the OperationalDeviceProxy and hence all the stuff we _think_ is cleaning up the CommissioneeDeviceProxy is doing nothing of the sort?